Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/DefaultPrimaryPointerSelector.cs

```diff
@@ -0,0 +1,156 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System.Collections.Generic;
+
+namespace Microsoft.MixedReality.Toolkit.Input
+{
+    /// <summary>
+    /// Default primary pointer selector. The primary pointer is chosen among all interaction enabled ones using the following rules in order:
+    ///   1. Currently pressed pointer that has been pressed for the longest
+    ///   2. Pointer that was released most recently
+    ///   3. Pointer that became interaction enabled most recently
+    /// </summary>
+    public class DefaultPrimaryPointerSelector : IMixedRealityPrimaryPointerSelector, IMixedRealityPointerHandler
+    {
+        private readonly Dictionary<IMixedRealityPointer, PointerInfo> pointerInfos = new Dictionary<IMixedRealityPointer, PointerInfo>();
+
+        public void Initialize()
+        {
+            IMixedRealityInputSystem inputSystem;
+            if (MixedRealityServiceRegistry.TryGetService<IMixedRealityInputSystem>(out inputSystem))
+            {
+                inputSystem.RegisterHandler<IMixedRealityPointerHandler>(this);
+            }
+        }
+
+        public void Destroy()
+        {
+            IMixedRealityInputSystem inputSystem;
+            if (MixedRealityServiceRegistry.TryGetService<IMixedRealityInputSystem>(out inputSystem))
+            {
+                inputSystem.UnregisterHandler<IMixedRealityPointerHandler>(this);
+            }
+        }
+
+        #region IMixedRealityPrimaryPointerSelector
+
+        public void RegisterPointer(IMixedRealityPointer pointer)
+        {
+            var pointerInfo = new PointerInfo(pointer);
+            pointerInfos.Add(pointer, pointerInfo);
+        }
+
+        public void UnregisterPointer(IMixedRealityPointer pointer)
+        {
+            pointerInfos.Remove(pointer);
+        }
+
+        virtual public IMixedRealityPointer Update()
+        {
+            IMixedRealityPointer primaryPointer = null;
+            PointerInfo primaryInfo = null;
+
+            foreach (var keyValue in pointerInfos)
```

Comment:
>foreach (var keyValue in pointerInfos) [](start = 12, length = 38)

Similar comment that I had in another review, just be wary that if this is called in every single Update() loop, that this isn't spawning additional GC garbage. It would be worth doing a quick profile (even in player) to see if this occurs, and if so, it might be worth snapshotting the list on register/unregister to eat the cost there instead of on each update.

Reply:
Good that you bring that up because we need to dispel that myth:

![image](https://user-images.githubusercontent.com/47451126/59858522-096a6300-9373-11e9-8026-884846e7f3c1.png)

Let's go over what happens in that [`foreach`](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/foreach-in):

1. [`Dictionary<TKey,TValue>.GetEnumerator`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.dictionary-2.getenumerator?view=netframework-4.8) is called, which returns a [`Dictionary<TKey,TValue>.Enumerator`](Dictionary<TKey,TValue>.Enumerator) struct, so no allocation there.
1. [`Dictionary<TKey,TValue>.Enumerator.MoveNext`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.dictionary-2.enumerator.movenext?view=netframework-4.8) is called to move to the first element.
1. [`Dictionary<TKey,TValue>.Enumerator.Current`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.dictionary-2.enumerator.current?view=netframework-4.8) is called which returns a [`KeyValuePair<TKey,TValue>`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.keyvaluepair-2?view=netframework-4.8) struct, again no allocation there.
1. Goto #2 until MoveNext returns false.

Does that sounds alright to you? Where did you think those allocations could be coming from?

Reply:
Awesome, thanks for doing this :)

Yes, I wasn't sure about this because when I was profiling some of the default pointer mediator stuff (which also did foreach over the collection) I found that it did have non-zero GC Alloc on basically every frame.

Reply:
Are you positive those were coming from a foreach? I'd like to have a look if so. At the moment I see DefaultPointerMediator.UpdatePointers() making extensive use of foreach but zero allocations.